internal/kconfig/kconfig.go

```diff
@@ -0,0 +1,249 @@
+package kconfig
+
+import (
+	"bufio"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/ebpf/btf"
+	"github.com/cilium/ebpf/internal"
+)
+
+// Find find a kconfig file on the host.
+// It first reads from /boot/config- of the current running kernel and tries
+// /proc/config.gz if nothing was found in /boot.
+// If none of the file provide a kconfig, it returns an error.
+func Find() (*os.File, error) {
+	kernelRelease, err := internal.KernelRelease()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get kernel release: %w", err)
+	}
+
+	path := "/boot/config-" + kernelRelease
+	f, err := os.Open(path)
+	if err == nil {
+		return f, nil
+	}
+
+	f, err = os.Open("/proc/config.gz")
+	if err == nil {
+		return f, nil
+	}
+
+	return nil, fmt.Errorf("neither %s nor /proc/config.gz provide a kconfig", path)
+}
+
+// Parse parses the kconfig file for which a reader is given.
+// All the CONFIG_* set will be put in the returned map as key with their
+// corresponding value as map value.
+// If the kconfig file is not valid, error will be returned.
+func Parse(source io.ReaderAt) (map[string]string, error) {
+	var r io.Reader
+	zr, err := gzip.NewReader(io.NewSectionReader(source, 0, math.MaxInt64))
+	if err != nil {
+		r = io.NewSectionReader(source, 0, math.MaxInt64)
+	} else {
+		// Source is gzip compressed, transparently decompress.
+		r = zr
+	}
+
+	ret := make(map[string]string)
+
+	s := bufio.NewScanner(r)
+	for s.Scan() {
+
+		line := s.Text()
+		err = processKconfigLine(line, ret)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse line: %w", err)
+		}
+	}
+
+	if err := s.Err(); err != nil {
+		return nil, fmt.Errorf("cannot parse: %w", err)
+	}
+
+	if zr != nil {
+		return ret, zr.Close()
+	}
+
+	return ret, nil
+}
+
+// Golang translation of libbpf bpf_object__process_kconfig_line():
+// https://github.com/libbpf/libbpf/blob/fbd60dbff51c870f5e80a17c4f2fd639eb80af90/src/libbpf.c#L1874
+// It does the same checks but does not put the data inside the BPF map.
+func processKconfigLine(line string, m map[string]string) error {
+	// Ignore empty lines and "# CONFIG_* is not set".
+	if !strings.HasPrefix(line, "CONFIG_") {
+		return nil
+	}
+
+	key, value, found := strings.Cut(line, "=")
+	if !found {
+		return fmt.Errorf("line %q does not contain separator '='", line)
+	}
+
+	if len(value) == 0 {
+		return fmt.Errorf("line %q has no value", line)
+	}
+
+	_, ok := m[key]
+	if !ok {
```

Comment:
Isn't this the opposite of what kconfig does? In my experience the last value wins.

Should we allow overwriting values at all?

Reply:
Ah weird. Can you add a comment to that effect?